### ISSUE TITLE
The Training Parameters for ENZYMES Dataset

### ISSUE THREAD

**Issue by chLWang** [davidbuterez/edge-set-attention#2]
Hello, I have read your paper and the code you provided. Your work is really excellent. I tried to reproduce your code on ENZYMES dataset, but the evaluation metrics I obtained are not as high as those reported in your paper. Could you please provide the training parameters you used on ENZYMES dataset? Thank you very much！

**Comment by davidbuterez** [comment]
Hi, and thanks for the interest in ESA! There are indeed some reproducibility considerations that I plan to document more thoroughly when I will have more free time. It is worth noting that I encountered different results for different PyTorch/CUDA versions, and I also think that one PyTorch Geometric update switched the rows of the `edge_index` tensor, which could affect the inner workings of ESA. Models that use dropout or batch normalization will also not be fully reproducible, as although seeds were set for Python/PyTorch/NumPy/CUDA, the full CUDA determinism flags were not set for performance reasons (also, most models do not use dropout/BN anyway). In this case, as well as some other instances, there are also a few hyperparameters that can influence the result, but are generally too specific or simple to be set from the command line, and must be altered manually for each dataset. For ENZYMES, it is better to use the default `eps` value in LayerNorm (instead of 1e-8) and use `k=64` for PMA. The GitHub version also uses a single `Linear` layer for the output layer, instead of an MLP (the MLP works better here).

All in all, with the following command:

```
python -m esa.train --dataset ENZYMES --dataset-download-dir <> --dataset-one-hot --dataset-target-name None --lr 0.0001 --batch-size 512 --norm-type LN --early-stopping-patience 100 --monitor-loss-name MCC --graph-dim 512 --apply-attention-on edge --use-mlps --mlp-hidden-size 32 --out-path <> --wandb-project-name <> --hidden-dims 1024 1024 1024 1024 1024 1024 1024 1024 1024 --num-heads 128 128 128 128 128 128 128 128 128 --sab-dropout 0 --mab-dropout 0 --pma-dropout 0 --seed 0 --optimiser-weight-decay 1e-10 --gradient-clip-val 0.5 --xformers-or-torch-attn xformers --mlp-type gated_mlp --use-bfloat16 --layer-types M M M M M M P S S --mlp-layers 2 --pma-residual-dropout 0.2 --use-mlp-ln yes --pre-or-post post --mlp-dropout 0 --attn-residual-dropout 0
```

This is the expected output:
```
wandb: Run summary:
wandb:                        Test AP 0.79227
wandb:                     Test AUROC 0.92836
wandb:                  Test Accuracy 0.77543
wandb:                        Test F1 0.78993
wandb:                       Test MCC 0.75518
```

This is an [edited version of the repo](https://www.dropbox.com/scl/fi/scp4uwshoq4mzj7hll775/edge-set-attention-main-ENZYMES.zip?rlkey=xvr2x44msdmqtf7e69m1t15pb&st=lrqyofi7&dl=0) for convenience, including the edits mentioned above, and switching the rows of the `edge_index`.

For reference, I ran this code in a fresh environment with the following versions:

```
mamba install pytorch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1 pytorch-cuda=12.4 -c pytorch -c nvidia -y;
pip install pytorch_lightning wandb bitsandbytes rdkit ogb scikit-learn pandas admin-torch madgrad yacs torch-optimi ninja packaging;
pip install xformers==0.0.29;
pip install flash-attn --no-build-isolation;
pip install torch_geometric;
pip install torch-scatter -f https://data.pyg.org/whl/torch-2.5.1+cu124.html;
pip install torch-sparse -f https://data.pyg.org/whl/torch-2.5.1+cu124.html;
pip install datasets transformers accelerate;
```
